### PR TITLE
Support registering Application Services when running with workers under Complement.

### PR DESCRIPTION
**Pull request by reivilibre**
This goes towards fixing #12825.

### CODE REVIEW

File: docker/configure_workers_and_start.py [hunk 1]
```diff
@@ -488,11 +491,23 @@ def generate_worker_files(
     master_log_config = generate_worker_log_config(environ, "master", data_dir)
     shared_config["log_config"] = master_log_config
 
+    # Find application service registrations
+    appservice_registrations = None
+    appservice_registration_dir = os.environ.get("SYNAPSE_AS_REGISTRATION_DIR")
+    if appservice_registration_dir:
+        # Scan for all YAML files that should be application service registrations.
+        appservice_registrations = [
+            str(reg_path.absolute())
```
Comment (clokep) [inline docker/configure_workers_and_start.py:500]: I can't find any documentation for `absolute()`. Kind of sounds like [`resolve()`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.resolve)?

File: docker/conf-workers/shared.yaml.j2 [hunk 2]
```diff
@@ -6,4 +6,13 @@
 redis:
     enabled: true
 
-{{ shared_worker_config }}
\ No newline at end of file
+## Application Services ##
+{% if appservice_registrations is not none %}
```
Comment (clokep) [inline docker/conf-workers/shared.yaml.j2:10]: Maybe?

```suggestion
{% if appservice_registrations is not none %}
## Application Services ##
```

File: docker/configure_workers_and_start.py (hunk 1 above)
Comment (reivilibre) [reply docker/configure_workers_and_start.py:500]: Huh, weird
```python
>>> from pathlib import Path
>>> p = Path("/tmp")
>>> p.absolute()
PosixPath('/tmp')
>>> p.__class__
<class 'pathlib.PosixPath'>
>>> help(p.absolute)
```

```
Help on method absolute in module pathlib:

absolute() method of pathlib.PosixPath instance
    Return an absolute version of this path.  This function works
    even if the path doesn't point to anything.
    
    No normalization is done, i.e. all '.' and '..' will be kept along.
    Use resolve() to get the canonical path to a file.
```

Nothing against using `resolve`, this just seemed more intuitive to use (no idea why it's not documented!)

**Comment by reivilibre** [comment]
> This exists on the non-worker version:
> 
> https://github.com/matrix-org/synapse/blob/66a3f2fb34d5552a3f6d45a8c1230e4bdfc080ae/docker/conf/homeserver.yaml#L164-L168

That's not the Complement one.

The non-docker one I looked at does some.. sed nastiness..

https://github.com/matrix-org/synapse/blob/4fef76ca348209b7c9dd3c17d5f3d8ef12623c1b/docker/complement/conf/homeserver.yaml#L108-L109
https://github.com/matrix-org/synapse/blob/4fef76ca348209b7c9dd3c17d5f3d8ef12623c1b/docker/complement/conf/start.sh#L8-L18
(sigh)

In the workers case, we have Jinja2 so I thought it would be nicer to use that instead.
Note we have to scan the directory for all the appservices, so I thought doing that in Python was a little less yucky than having a space-separated env variable or something like that.

File: docker/configure_workers_and_start.py (hunk 1 above)
Comment (clokep) [reply docker/configure_workers_and_start.py]: It is likely fine, was just confused. 🤷

**Review by clokep — APPROVED** [review]

**Pull request merged**